pkg/controller/dynamicupgradeableconditionscontroller.go

```diff
@@ -0,0 +1,245 @@
+	c := &dynamicUpgradeableConditionController{
+		name:                   name,
+		clusterExtensionClient: clusterExtensionClient,
+		kubeclient:             kubeclient,
+		operatorClient:         operatorClient,
+		prefixes:               prefixes,
+		logger:                 klog.FromContext(context.TODO()).WithName(name),
+	}
+
+	return factory.New().WithSync(c.sync).WithSyncDegradedOnError(operatorClient).WithInformers(operatorClient.Informer(), clusterExtensionClient.Informer().Informer()).ToController(name, eventRecorder)
+}
+
+func (c *dynamicUpgradeableConditionController) sync(ctx context.Context, _ factory.SyncContext) error {
+	c.logger.V(4).Info("sync started")
+	defer c.logger.V(4).Info("done syncing")
+
+	updateStatusFuncs, err := c.ensureControllersManaged()
+	if err != nil {
+		c.logger.V(4).Error(err, "error verifying OLM state")
+		return err
+	}
+
+	incompatibleOperatorsUpdateStatusFuncs, err := c.checkIncompatibleOperators()
+	if err != nil {
+		c.logger.V(4).Error(err, "error checking incompatible operators")
+		return err
+	}
+
+	updateStatusFuncs = append(updateStatusFuncs, incompatibleOperatorsUpdateStatusFuncs...)
+	if _, _, updateErr := v1helpers.UpdateStatus(ctx, c.operatorClient, updateStatusFuncs...); updateErr != nil {
+		c.logger.V(4).Info(fmt.Sprintf("Error updating operator condition status: %v", updateErr))
+		return updateErr
+	}
+	return nil
+}
+
+func (c *dynamicUpgradeableConditionController) ensureControllersManaged() ([]v1helpers.UpdateStatusFunc, error) {
+	c.logger.V(4).Info("verifying state of olm controllers")
+	defer c.logger.V(4).Info("done verifying state of olm controllers")
+	opSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return nil, err
+	}
+	if opSpec.ManagementState != operatorv1.Managed {
+		return nil, nil
+	}
+
+	updateStatusFuncs := make([]v1helpers.UpdateStatusFunc, 0, len(c.prefixes))
+	for _, prefix := range c.prefixes {
+		updateStatusFuncs = append(updateStatusFuncs, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:   fmt.Sprintf("%sUpgradeable", prefix),
+			Status: operatorv1.ConditionTrue,
+		}))
+	}
+	return updateStatusFuncs, nil
+}
+
+func (c *dynamicUpgradeableConditionController) checkIncompatibleOperators() ([]v1helpers.UpdateStatusFunc, error) {
+	// Get the raw version from the releaseEnvVar environment variable
+	raw, ok := os.LookupEnv(releaseEnvVar)
+	if !ok || raw == "" {
+		// No env var set, try again later
+		return nil, nil
+	}
+
+	current, err := semver.ParseTolerant(raw)
+	if err != nil {
+		return nil, fmt.Errorf("cluster version has invalid desired release version: %w", err)
+	}
+
+	next, err := nextY(current)
+	if err != nil {
+		c.logger.V(4).Error(err, "Error finding next OCP Y-stream release")
+		return nil, err
+	}
```

Comment:
None of this changes unless the process restarts, right? I'd recommend two things:
1. Do envvar lookups further up the call stack, maybe in `runOperator` in `main.go`?
2. Make a helper function that accepts `raw` and returns `next, err` and also call _that_ function in `main.go`, passing `next` into the code that checks for incompatible operators.

Reply:
I just found that we already have the payload version via `OPERATOR_IMAGE_VERSION`, and library-go has a function that we already use that abstracts it. Look for `status.VersionForOperatorFromEnv()` in `main.go`

We can probably piggyback on all of that. Editing my comment above with this new info.